### ISSUE TITLE
No movement with no headset

### ISSUE THREAD

**Issue by SushilRagoonath** [bjornbytes/lovr#359]
the headset driver is handling per frame input updates but if we use no headset then there is no movement / camera movement, the solution would be using the desktop driver but then there is stereo mode. Is there any way to use the desktop vr driver with regular mono view?

**Comment by SushilRagoonath** [comment]
I see that that you can create a fps controller but having the mono option could remove the need for it.

**Comment by mcclure** [comment]
In my [fork of lovr](https://github.com/mcclure/lovr) ("andi" branch) I use a patch like this to give the desktop driver a mono mode:

```
diff --git a/src/modules/headset/headset_desktop.c b/src/modules/headset/headset_desktop.c
--- a/src/modules/headset/headset_desktop.c
+++ b/src/modules/headset/headset_desktop.c
@@ -93,6 +106,8 @@
 }
 
 static bool desktop_getViewAngles(uint32_t view, float* left, float* right, float* up, float* down) {
+  *left = -state.fov * .5f;
+  *right = state.fov * .5f;
   float aspect;
   uint32_t width, height;
   desktop_getDisplayDimensions(&width, &height);
@@ -173,10 +195,14 @@
 }
 
 static void desktop_renderTo(void (*callback)(void*), void* userdata) {
+  bool stereo = false;
+  uint32_t width, height;
+  desktop_getDisplayDimensions(&width, &height);
+  float aspect = (float) width / (float) height / (1 + stereo);
   float left, right, up, down;
   desktop_getViewAngles(0, &left, &right, &up, &down);
-  Camera camera = { .canvas = NULL, .viewMatrix = { MAT4_IDENTITY }, .stereo = true };
-  mat4_fov(camera.projection[0], tanf(left), tanf(right), tanf(up), tanf(down), state.clipNear, state.clipFar);
+  Camera camera = { .canvas = NULL, .viewMatrix = { MAT4_IDENTITY }, .stereo = stereo };
+  mat4_fov(camera.projection[0], tanf(left) * aspect, tanf(right) * aspect, tanf(up), tanf(down), state.clipNear, state.clipFar);
   mat4_multiply(camera.viewMatrix[0], state.headTransform);
   mat4_invert(camera.viewMatrix[0]);
   mat4_set(camera.projection[1], camera.projection[0]);
```

I think it would be great if this were something you could control in conf.lua.

**Comment by SushilRagoonath** [comment]
Thanks,that makes sense.I think I'll leave this open just a bit longer in-case someone else has something interesting but I may end up doing something like this.

**Comment by bjornbytes** [comment]
The spirit of the driver is to simulate a stereo VR headset and provide an implementation of the lovr.headset module.  Its purpose is not to provide keyboard and mouse controls for debugging or for implementing pancake apps, although it's commonly used for that.

What are the reasons to have the desktop headset driver use stereo rendering?

In the past this meant calling lovr.draw twice, although with multiview that's not really a thing.  It can sometimes be useful to have two views separated by a virtual IPD to test out stereo-only techniques, for example combined-eye frustum culling or testing stereo post-processing shaders.

Also, as of last version the headset module is no longer locked to having 2 views.  There can be any number of views, so the idea that "the desktop driver is stereo so that it matches headsets" doesn't hold as much weight -- portable projects potentially need to be able to handle different view counts anyway.

The view count of a driver is also separate from what the mirror window renders.  `lovr.mirror` can just as easily do `lovr.graphics.fill` with custom uvs to crop out a single eye to display.  Now, this is not possible right now specifically with the desktop driver because it doesn't render to a texture -- it renders directly to the window.  But that could be changed to make this possible.

Adding a flag that results in making things mono in this case is up for consideration, but I would prefer if it were generic instead of something that applies to a single driver.  For example, if I were to say "hey lovr.headset please use mono rendering", I would expect this to also apply to something like the OpenXR view configuration, where you can actually request that the XR system use mono rendering (it's for phone AR, requesting mono would fail on a stereo headset).

I can't generally recommend using the desktop driver as "a way to get keyboard controls".

TLDR idk why this got so long

- Should the desktop headset driver simulate one view instead of two views?  Testing out stereo techniques and/or compatibility with stereo headsets no longer seems like a compelling-enough reason to keep it stereo, and most people don't like the stereoness.
- Should the desktop headset driver render to a texture so people can use lovr.mirror to make it mono?
- Should conf.lua provide an option to explicitly set the view configuration

**Comment by mcclure** [comment]
> Adding a flag that results in making things mono in this case is up for consideration… I would expect this to also apply to something like the OpenXR view configuration

There is also a settable flag along these lines in Oculus Rift driver.

> Testing out stereo techniques and/or compatibility with stereo headsets no longer seems like a compelling-enough reason to keep it stereo

I think testing stereo techniques *is* useful (what if I'm testing shaders?) but it would be neat if this could be switched off.

In general if I were shipping a game I would like the mirror window to display a mono view of some kind, it just more matches what I think a user would expect to see if they looked at the screen.

**Comment by SushilRagoonath** [comment]
I like the desktop driver going to a render texture, would still keep the idea of headsets being vr but allowing very easy mono support.

**Comment by jmiskovic** [comment]
I use this bit of code on dev machine to have mono rendering with camera controls still working:

```lua
if lovr.headset.getName() == 'Simulator' then
  lovr.mirror = function ()
    lovr.graphics.clear()
    lovr.graphics.transform(mat4(lovr.headset.getViewPose(1)):invert())
    lovr.draw()
  end
end
```

**Comment by SushilRagoonath** [comment]
Thanks this works exactly as I wanted. For future readers you need to override `lovr.mirror` with the above.
